eventqueue/processor.go

```diff
@@ -0,0 +1,229 @@
+
+	return nil
+}
+
+// Stop the processor.
+func (p *Processor[T]) Close() error {
+	if !p.stopped.CompareAndSwap(false, true) {
+		// Already stopped
+		return nil
+	}
+
+	// Send a signal to stop
+	close(p.stopCh)
+
+	// Blocks until process loop ends
+	t := time.NewTimer(5 * time.Second)
+	select {
+	case p.processorRunningCh <- struct{}{}:
+		// All good
+		if !t.Stop() {
+			<-t.C
+		}
+	case <-t.C:
+		return errors.New("Processor loop did not stop in 5s")
+	}
+
+	return nil
+}
+
+// Start the processing loop if it's not already running.
+// This must be invoked while the caller has a lock.
+func (p *Processor[T]) process(isNext bool) {
+	// Do not start a loop if it's already running
+	select {
+	case p.processorRunningCh <- struct{}{}:
+		// Nop - fallthrough
+	default:
+		// Already running
+		if isNext {
+			// If this is the next item, send a reset signal
+			// Use a select in case another goroutine is sending a reset signal too
+			select {
+			case p.resetCh <- struct{}{}:
+			default:
+			}
+		}
+		return
+	}
+
+	go p.processLoop()
+}
+
+// Processing loop.
+func (p *Processor[T]) processLoop() {
+	defer func() {
+		// Release the channel when exiting
+		<-p.processorRunningCh
+	}()
+
+	var (
+		r             T
+		ok            bool
+		t             kclock.Timer
+		scheduledTime time.Time
+		deadline      time.Duration
+	)
+
+	for {
+		// Continue processing items until the queue is empty
+		p.queueLock.Lock()
+		r, ok = p.queue.Peek()
+		p.queueLock.Unlock()
+		if !ok {
+			return
+		}
+
+		// Check if after obtaining the lock we have a stop or reset signals
+		// Do this before we create a timer
+		select {
+		case <-p.stopCh:
+			// Exit on stop signals
+			return
+		case <-p.resetCh:
+			// Restart the loop on reset signals
+			continue
+		default:
+			// Nop, proceed
+		}
+
+		scheduledTime = r.ScheduledTime()
+		deadline = scheduledTime.Sub(p.clock.Now())
+
+		// If the deadline is less than 0.5ms away, execute it right away
+		// This is more efficient than creating a timer
+		if deadline < 500*time.Microsecond {
+			p.execute(r)
+			continue
+		}
+
+		t = p.clock.NewTimer(deadline)
+		select {
+		// Wait for when it's time to execute the item
+		case <-t.C():
+			p.execute(r)
+
+		// If we get a reset signal, restart the loop
+		case <-p.resetCh:
+			// Restart the loop
+			continue
+
+		// If we receive a stop signal, exit
+		case <-p.stopCh:
+			// Stop the timer and exit the loop
+			if !t.Stop() {
+				<-t.C()
+			}
+			return
+		}
+	}
+}
+
+// Executes a item when it's time.
+func (p *Processor[T]) execute(r T) {
+	// Pop the item now that we're ready to process it
+	// There's a small chance this is a different item than the one we peeked before
+	p.queueLock.Lock()
+	// For safety, let's peek at the first item before popping it and make sure it's the same object
+	// It's unlikely, but if it's a different object then restart the loop
+	peek, ok := p.queue.Peek()
+	if !ok || peek != r {
+		p.queueLock.Unlock()
```

Comment:
I'd rather have explicit controls on the locks here so I can release them as soon as they're not needed anymore, since they block every other operation.

Reply:
Hm, defer is doing the exact same though right? It is executed exactly after we return except for the last case where we spawn a go routine. We are not saving anything by not using defer, and it makes things a lot harder to read/understand what the code is doing by being non-idiomatic.